comp/metadata/packagesigning/packagesigningimpl/decryptgpg.go

```diff
@@ -0,0 +1,121 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package packagesigningimpl
+
+import (
+	"io"
+	"net/http"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/DataDog/datadog-agent/comp/core/log"
+	pkgUtils "github.com/DataDog/datadog-agent/comp/metadata/packagesigning/utils"
+	pgp "github.com/ProtonMail/go-crypto/openpgp"
+)
+
+// SigningKey represents relevant fields for a package signature key
+type SigningKey struct {
+	Fingerprint    string                  `json:"signing_key_fingerprint"`
+	ExpirationDate string                  `json:"signing_key_expiration_date"`
+	KeyType        string                  `json:"signing_key_type"`
+	Repositories   []pkgUtils.Repositories `json:"repositories"`
+}
+
+type repoFile struct {
+	filename     string
+	repositories []pkgUtils.Repositories
+}
+
+const (
+	noExpDate  = "9999-12-31"
+	formatDate = "2006-01-02"
+)
+
+// decryptGPGFile parse a gpg file (local or http) and extract signing keys information
+// Some files can contain a list of repositories.
+// It is called twice to parse the keyring as armored keyring or not
+func decryptGPGFile(allKeys map[string]SigningKey, gpgFile repoFile, keyType string, client *http.Client, logger log.Component) {
+	err := decrypt(allKeys, gpgFile, false, keyType, client)
```

Comment:
It looks `decrypt` could fail not only because of the armored condition, but still we blindly call it if it fails the fist time. 

It looks like we could benefit from extracting the reading portion from decrypt into separate method, so we can isolate reading errors before calling `decryptGPGReader`

Pseudo code

```go
func decryptGPGFile(allKeys map[string]SigningKey, gpgFile repoFile, keyType string, client *http.Client, logger log.Component) {
	reader, closeFunc, err := readGPGFile(client, gpgFile)

	if err != nil {
		logger.Infof("Error while reading gpg file %s: %s", gpgFile.filename, err)
		return
	}

	defer closeFunc()

	err = decryptGPGReader(allKeys, reader, false, keyType, gpgFile.repositories)
	if err != nil {
		err = decryptGPGReader(allKeys, reader, true, keyType, gpgFile.repositories)
		if err != nil {
			logger.Infof("Error while parsing gpg file %s: %s", gpgFile.filename, err)
		}
	}
}

func readGPGFile(client *http.Client, gpgFile repoFile) (io.Reader, close func(), error) {
	var reader io.Reader
	if strings.HasPrefix(gpgFile.filename, "http") {
		response, err := client.Get(gpgFile.filename)
		if err != nil {
			return reader, func() {}, err
		}
		return response.Body, func() { response.Body.Close() },  nil
	} else {
		file, err := os.Open(strings.Replace(gpgFile.filename, "file://", "", 1))
		if err != nil {
			return reader, func() {},err
		}
		return file, func() { file.Close() },  nil
	}
}
```

Reply:
Yes I can probably split once more the method, thanks for the suggestion

Reply:
hum, I'm not sure that works. I did split like that because if I use twice the reader it's empty on the second call. So I need to read the file twice. With your suggestion the reader will point to the end of the buffer after the first call, I think